test-case-reporting/src/db_interfacing.ts

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+
+import {Build, DB, Job, TestCase, TestRun} from 'test-case-reporting';
+import {Database, TIMESTAMP_PRECISION} from './db';
+
+// Types in the DB namespace interface use snake_case instead of camelCase.
+/* eslint @typescript-eslint/camelcase: "off" */
+function getDateFromTimestamp(timestamp: number): Date {
+  const msConversionConstant: number = Math.pow(10, 3 - TIMESTAMP_PRECISION);
+  return new Date(timestamp * msConversionConstant);
+}
+
+export class TestResultsRecords {
+  constructor(private db: Database) {}
+
+  private async getDbBuild(buildNumber: string): Promise<DB.Build> {
+    return this.db<DB.Build>('builds')
+      .where('build_number', buildNumber)
+      .first();
+  }
+
+  private async getBuildWithEmptyJobs(buildNumber: string): Promise<Build> {
+    const dbBuild = await this.getDbBuild(buildNumber);
+    return {
+      commitSha: dbBuild.commit_sha,
+      buildNumber: dbBuild.build_number,
+      startedAt: getDateFromTimestamp(dbBuild.started_at),
+      jobs: [],
+    };
+  }
+
+  private makeTestCaseFromRow(row): TestCase {
+    return {
+      name: row.test_case_name,
+      createdAt: getDateFromTimestamp(row.test_case_created_at),
+    };
+  }
+
+  private makeTestRunFromRow(row): TestRun {
+    return {
+      testCase: this.makeTestCaseFromRow(row),
+      status: row.test_run_status,
+      timestamp: getDateFromTimestamp(row.test_run_timestamp),
+      durationMs: row.duration_ms,
+    };
+  }
+
+  // I need help typechecking this
+  /**
+   * Helper method.
+   * Takes a row of the join of every table.
+   * Makes a job containing only the test run described in that row.
+   * @param row A row of the join of every table, with some columns
+   * renamed for clarity.
+   */
+  private makeJobFromRow(row): Job {
+    return {
+      jobNumber: row.job_number,
+      testSuiteType: row.test_suite_type,
+      startedAt: getDateFromTimestamp(row.job_started_at),
+      testRuns: [this.makeTestRunFromRow(row)],
+    };
+  }
+
+  async getBuild(buildNumber: string): Promise<Build> {
+    const build = await this.getBuildWithEmptyJobs(buildNumber);
+    const joinOfEveryTable = await this.db('builds')
+      .select([
+        'test_runs.job_id',
+        'jobs.started_at as job_started_at',
+        'jobs.test_suite_type',
+        'jobs.job_number',
+        'jobs.job_number',
+
+        'test_cases.name as test_case_name',
+        'test_cases.created_at as test_case_created_at',
+
+        'test_runs.status as test_run_status',
+        'test_runs.timestamp as test_run_timestamp',
+        'test_runs.duration_ms as test_run_duration_ms',
+      ])
+      .where('build_number', buildNumber)
+      .join('jobs', 'jobs.build_id', 'builds.id')
+      .join('test_runs', 'test_runs.job_id', 'jobs.id')
+      .join('test_cases', 'test_cases.id', 'test_runs.test_case_id');
+
+    const jobs = joinOfEveryTable
```

Comment:
While I tend to like functional solutions, I worry this is introducing a lot of complexity to shoehorn the DB row into the ideal type we wanted. The logic could get more complex for the history of a single test, which will also mean grouping by build in addition to job. What do you think about removing the ''jobs'' and ''testRuns'' properties entirely from the Build and Job types, keeping them very flat, and maybe letting one object whole info for the build, job, and test?

Reply:
I like that idea; it'll simplify things.

Reply:
Done.